### ISSUE TITLE
What is the desired output for _select_targets in common.py?

### ISSUE THREAD

**Issue by tsKenneth** [pytorch/captum#288]
Hi again,

I have a question to ask about the _select_targets function, specifically when used for the DeepLift implementation. I figured out that the output passed into this function is based on the output from the last layer of the architecture. For my architecture, my last layer is a log_softmax. Sorry if it is a silly question but should i return the predicted class (only 2 classes), the loss value or the class probability of the target class as output?

Attached below is the code snippet for _select_targets for your reference.

```
def _select_targets(output, target):
    output = output[0]
    num_examples = output.shape[0]
    dims = len(output.shape)

    if target is None:
        return output
    elif isinstance(target, int) or isinstance(target, tuple):
        return _verify_select_column(output, target)
    elif isinstance(target, torch.Tensor):
        if torch.numel(target) == 1 and isinstance(target.item(), int):
            return _verify_select_column(output, target.item())
        elif len(target.shape) == 1 and torch.numel(target) == num_examples:
            assert dims == 2, "Output must be 2D to select tensor of targets."
            return torch.gather(output, 1, target.reshape(len(output), 1))
        else:
            raise AssertionError(
                "Tensor target dimension %r is not valid." % (target.shape,)
            )
    elif isinstance(target, list):
        assert len(target) == num_examples, "Target list length does not match output!"
        if type(target[0]) is int:
            assert dims == 2, "Output must be 2D to select tensor of targets."
            return torch.gather(output, 1, torch.tensor(target).reshape(len(output), 1))
        elif type(target[0]) is tuple:
            return torch.stack(
                [output[(i,) + targ_elem] for i, targ_elem in enumerate(target)]
            )
        else:
            raise AssertionError("Target element type in list is not valid.")
    else:
        raise AssertionError("Target type %r is not valid." % target)
```

**Comment by vivekmig** [comment]
Hi @tsKenneth, the purpose of this _select_targets method is to select a single (scalar) value for each example in the output of your model to compute attributions based on the given target parameter. Each attribution method answers the question of how important each input value was towards a particular output value.  

More concretely, in your case it seems your output is in the shape N x 2, with N being the number of examples and 2 being the number of classes. The purpose of targets is to select one of the 2 values for each example. You could then pass targets as just a single integer 0 or 1, which corresponds to attributing to the output for class 0 or class 1 respectively for all examples. If you are passing a batch of multiple examples (e.g. N = 4), you can also pass a different target for each example, by providing targets as a list or tensor, e.g. [0, 1, 0, 0], which would return attributions for class 0 for the first, third and fourth input examples (how important each input value was for the prediction of class 0), and class 1 for the second one (how important each input value was for the prediction of class 1). 

Regarding your question about what model output to use, theoretically, any scalar output can be used to compute attributions. The best option might be to use the model output before applying softmax (logits), see section 3.6 of the DeepLift paper or #91 from one of the authors for some reasoning why.

**Comment by tsKenneth** [comment]
Thank you for the clarification!
